tendermint/src/abci/data.rs

```diff
@@ -1,64 +1,78 @@
-use crate::{Error, Kind};
-use serde::{Deserialize, Deserializer, Serialize, Serializer};
-use std::{
-    fmt::{self, Display},
-    str::FromStr,
-};
-use subtle_encoding::hex;
+use crate::Error;
+use serde::{Deserialize, Serialize};
+use std::convert::{TryFrom, TryInto};
+use tendermint_proto::serializers::bytes::Base64;
 
 /// ABCI transaction data.
 ///
 /// Transactions are opaque binary blobs which are validated according to
 /// application-specific rules.
-#[derive(Clone, Debug, Eq, PartialEq, Default)]
+#[derive(Clone, Debug, Eq, PartialEq, Default, Serialize, Deserialize)]
+#[serde(try_from = "Base64", into = "Base64")]
 pub struct Data(Vec<u8>);
 
-impl Data {
-    /// Borrow the data as bytes
-    pub fn as_bytes(&self) -> &[u8] {
-        self.0.as_ref()
-    }
-}
+impl TryFrom<Vec<u8>> for Data {
+    type Error = Error;
 
-impl AsRef<[u8]> for Data {
-    fn as_ref(&self) -> &[u8] {
-        self.0.as_ref()
+    fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
+        // Todo: Are there any restrictions on the incoming data?
+        Ok(Self(value))
     }
 }
 
-impl Display for Data {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        for byte in &self.0 {
-            write!(f, "{:02X}", byte)?;
-        }
-        Ok(())
+impl From<Data> for Vec<u8> {
+    fn from(value: Data) -> Self {
+        value.0
     }
 }
 
-impl FromStr for Data {
-    type Err = Error;
+impl TryFrom<Base64> for Data {
+    type Error = Error;
 
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        // Accept either upper or lower case hex
-        let bytes = hex::decode_upper(s)
-            .or_else(|_| hex::decode(s))
-            .map_err(|_| Kind::Parse)?;
+    fn try_from(value: Base64) -> Result<Self, Self::Error> {
+        value.0.try_into()
+    }
+}
 
-        Ok(Data(bytes))
+impl From<Data> for Base64 {
+    fn from(value: Data) -> Self {
+        Self(value.into())
     }
 }
 
-impl<'de> Deserialize<'de> for Data {
-    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
-        use serde::de::Error;
-        String::deserialize(deserializer)?
-            .parse()
-            .map_err(D::Error::custom)
+impl AsRef<[u8]> for Data {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
     }
 }
```

Comment:
Does this trait implementation give us any other functionality beyond keeping the `as_bytes()` method I recommendation above?